compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionUtils.cpp

```diff
@@ -56,6 +63,118 @@ static bool canInsOperandTieWithOutsOperand(OpOperand *insOperand) {
+  // TODO: the condition may be relaxed to support transpose or reduction on an
+  // arbirary dimension.
+  auto output = linalgOp.getOutputOperand(0);
+  auto outputIndexingMap = linalgOp.getTiedIndexingMap(output);
+  for (const auto &en : llvm::enumerate(outputIndexingMap.getResults())) {
+    auto expr = en.value();
+    if (auto dim = expr.dyn_cast<AffineDimExpr>()) {
+      if (dim.getPosition() != en.index()) return false;
+    } else {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/// Check if the op is elementwise and the output indexing map is identity.
+static bool isSimpleElementwise(linalg::LinalgOp op) {
+  if (op.getNumLoops() != op.getNumParallelLoops()) return false;
+
+  if (!allIndexingsAreProjectedPermutation(op)) return false;
+
+  for (OpOperand *opOperand : op.getOutputOperands()) {
+    if (!op.getTiedIndexingMap(opOperand).isIdentity()) return false;
+  }
+  return true;
+}
+
+/// Check if the use of operand is a reduction -> broadcast -> elementwise.
+/// An example:
+///   s = sum(a: <16x32xf32>) -> <16xf32>
+///   d = div(a: <16x32xf32>, broadcast(s to <16x32xf32>)) -> <16x32xf32>
+static bool isReductionBroadcastElementwise(OpOperand *operand) {
+  auto producer = operand->get().getDefiningOp<linalg::LinalgOp>();
+  auto consumer = dyn_cast<linalg::LinalgOp>(operand->getOwner());
+  if (!producer || !consumer) return false;
+
+  // Check if the producer is a simple reduction.
+  if (!isReductionOnInnermostDims(producer)) return false;
+
+  // Check if the reduction is broadcasted back for the elementwise op.
+  // TODO: We may need to relax the condition to support some broadcast with
+  // a unit dimension, e.g., <16x8xf32> -> <16xf32> -> <16x1x8xf32>.
+  auto producerResult = operand->get().cast<OpResult>();
+  auto outputIndexingMap = producer.getTiedIndexingMapForResult(producerResult);
+  auto inputIndexingMap = consumer.getTiedIndexingMap(operand);
+  if (outputIndexingMap != inputIndexingMap) return false;
+
+  // Check if the consumer is an elementwise with identity output indexing map.
+  if (!isSimpleElementwise(consumer)) return false;
+
+  // Check the input and output shapes are compatible. They are compatible when
+  //   1. the shapes are identical, or
+  //   2. the broadcasted input shape is the same as the output shape.
+  auto numInputs = producer.getNumInputs();
+  auto ewOutputType = consumer.getOutputOperand(0)->get().getType();
+  if (numInputs == 1) {
+    if (producer.getInputOperand(0)->get().getType() != ewOutputType)
+      return false;
+  } else {
+    assert(numInputs == 2 && "Expected two inputs to reduction");
+
+    // For a binary reduction, at least one of them should be in a full
+    // dimension. Here we put another restriction that the full input does not
+    // have a transpose, which may be relaxed later. For the other operand, we
+    // expect it to be broadcasted to the output shape.
+    OpOperand *fullInput = nullptr;
```

Comment:
Not great to use `nullptr`. If you want to, use `Optional<OpOperand *>` .

Reply:
Done. Thanks